api/rosetta/metadata.go

```diff
@@ -0,0 +1,100 @@
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rosetta
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/labstack/echo/v4"
+
+	"github.com/onflow/flow-go-sdk"
+
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+// MetadataRequest implements the request schema for /construction/metadata.
+// `Options` object in this request is generated by a call to `/construction/preprocess`,
+// and should be sent unaltered as returned by that endpoint.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#request-3
+type MetadataRequest struct {
+	NetworkID identifier.Network `json:"network_identifier"`
+	Options   object.Options     `json:"options"`
+}
+
+// MetadataResponse implements the response schema for /construction/metadata.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#response-3
+type MetadataResponse struct {
+	Metadata object.Metadata `json:"metadata"`
+}
+
+// Metadata implements the /construction/metadata endpoint of the Rosetta Construction API.
+// Metadata endpoint returns information required for constructing the transaction.
+// For Flow, that information includes the reference block and sequence number. Reference block
+// is the last indexed blcok, and is used to track transaction expiration. Sequence number is
+// the proposer account's public key sequence number. Sequence number is incremented for each
+// transaction and is used to prevent replay attacks.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#constructionmetadata
+func (c *Construction) Metadata(ctx echo.Context) error {
+
+	var req MetadataRequest
+	err := ctx.Bind(&req)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidEncoding(invalidJSON, err))
+	}
+
+	if req.NetworkID.Blockchain == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockchainEmpty))
+	}
+	if req.NetworkID.Network == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(networkEmpty))
+	}
+
+	if req.Options.AccountID.Address == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(addressEmpty))
+	}
+	if len(req.Options.AccountID.Address) != hexAddressSize {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(addressLength,
+			withDetail("have_length", len(req.Options.AccountID.Address)),
+			withDetail("want_length", hexAddressSize),
+		))
+	}
+
+	current, _, err := c.retriever.Current()
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, internal(txReferenceRetrieval, err))
```

Comment:
This I don't understand. The error is _not_ that you were not able to retrieve the reference block for a transaction. The error is purely on the `retriever.Current()`, so it should be either `blockRetrieval` or `currentRetrieval`. You are mixing roles between components here by putting more meaning into it than there is.

Reply:
While it's nice to have digestible error messages for user-facing APIs that they can understand, this could undermine our ability to efficiently help our users if they encounter issues, as they might come back with non descriptive generic errors instead of the specific one we'd need to be able to figure out what is the underlying problem.

I feel like in cases like this, the more precise and specific, the better. However, if we want to also show an error message that users can make sense of alongside it, I am not at all against it, especially since Rosetta allows us to get pretty creative with our errors.

Reply:
I agree with that sentiment, but I don't see how it applies in this case. If a user is trying to get meta data for a transaction, and it fails with `could not retrieve reference block`, there is no ambiguity at all. There is only one reference block within the Flow system, and it's clear that it's related to creating a transaction.